tokio/tests/async_send_sync.rs

```diff
@@ -265,6 +267,17 @@ async_assert_fn!(tokio::sync::watch::Sender<u8>::closed(_): Send & Sync);
 async_assert_fn!(tokio::sync::watch::Sender<Cell<u8>>::closed(_): !Send & !Sync);
 async_assert_fn!(tokio::sync::watch::Sender<Rc<u8>>::closed(_): !Send & !Sync);
 
+async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
+    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & !Sync);
```

Comment:
Why is it not `Sync`? It's a future — those can always be `Sync`.

Reply:
Why does the assertion fail if I use `Sync`?

```
error[E0277]: `dyn Future<Output = u8> + Send` cannot be shared between threads safely
   --> tokio/tests/async_send_sync.rs:114:26
    |
30  |   fn require_sync<T: Sync>(_t: &T) {}
    |                      ---- required by this bound in `require_sync`
...
114 |               require_sync(&f);
    |                            ^^ `dyn Future<Output = u8> + Send` cannot be shared between threads safely
...
270 | / async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
271 | |     _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & Sync);
```

Reply:
The assertion fails because the future returned by the method is not `Sync`. Mostly I'm thinking that it _should_ be `Sync`.

Reply:
The tests only pass in its current form.

Reply:
Yes, I know that. My point is that there is likely some change you could make to `get_or_init` that _would_ make the test pass with `Sync`.

Reply:
Why do we want the method to be `Sync`? Isn't it sufficient that OnceCell is `Sync`? Once we got the reference to an `OnceCell` instance we can still call `get_or_init`, can`t we?

Reply:
Generally it is more convenient for the user if the things implement as many traits as possible. For example, it not being `Sync` would prevent you from putting an async fn that calls `get_or_init` into an `Box<dyn Future + Send + Sync>`, even though there is no safety-concern with doing so.

Reply:
Thinking more about it, it is fine if we have both of these tests:
```Rust
async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & !Sync);
async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send + Sync>>): Send & Sync);
```

Reply:
> Generally it is more convenient for the user if the things implement as many traits as possible. For example, it not being `Sync` would prevent you from putting an async fn that calls `get_or_init` into an `Box<dyn Future + Send + Sync>`, even though there is no safety-concern with doing so.

That makes sense in general, but in our case the following works e.g.:

```
use std::future::Future;
use std::pin::Pin;
use tokio;
use tokio::sync::OnceCell;

static ONCE: OnceCell<u32> = OnceCell::new();

async fn some_computation() -> u32 {
    5
}

async fn call_get_or_init() -> u32 {
    *ONCE.get_or_init(|| some_computation()).await
}

async fn call_f(f: Pin<Box<dyn Future<Output = u32> + Send + Sync>>) {
    f.await;
}

#[tokio::main]
async fn main() {
    tokio::spawn(async {
        call_f(Box::pin(call_get_or_init())).await;
    });
}
```

Reply:
Yes. I realized it was because the future type itself is not `Sync` in the test. Your snippet is in the _second_ case of my two test lines.